.github/workflows/build-test.yaml

```diff
@@ -1,29 +1,246 @@
+          path: /tmp/.build-cache
+          key: ${{ runner.os }}-docker-${{ github.sha }}
+
+      - name: Build Docker image
+        uses: docker/build-push-action@v2
+        with:
+          context: .
+          push: false
+          file: docker/Dockerfile
+          cache-to: type=local,dest=/tmp/.build-cache,mode=max
+
+  unit-tests:
+    name: Run unit tests
+    needs:
+      - build
+    runs-on: ubuntu-latest
+    steps:
+      - name: Repository checkout
+        uses: actions/checkout@v2
+
+      - name: Setup Docker buildx
+        uses: docker/setup-buildx-action@v1
+
+      - name: Use cache
+        uses: actions/cache@v2
+        with:
+          path: /tmp/.build-cache
+          key: ${{ runner.os }}-docker-${{ github.sha }}
+
+      - name: Build Docker image
+        uses: docker/build-push-action@v2
+        with:
+          context: .
+          push: false
+          tags: debos-builder
+          file: docker/Dockerfile
+          cache-from: type=local,src=/tmp/.build-cache
+          load: true
+          target: builder
 
       - name: Run unit tests
         run: |
           docker-compose -f docker/unit-tests.test.yml \
-            up --build --exit-code-from=sut
+            up --exit-code-from=sut
+
+  recipe-tests:
+    name: Run recipe tests
+    needs:
+      - build
+    runs-on: ubuntu-latest
+    steps:
+      - name: Repository checkout
+        uses: actions/checkout@v2
+
+      - name: Setup Docker buildx
+        uses: docker/setup-buildx-action@v1
+
+      - name: Use cache
+        uses: actions/cache@v2
+        with:
+          path: /tmp/.build-cache
+          key: ${{ runner.os }}-docker-${{ github.sha }}
+
+      - name: Build Docker image
+        uses: docker/build-push-action@v2
+        with:
+          context: .
+          push: false
+          tags: debos
+          file: docker/Dockerfile
+          cache-from: type=local,src=/tmp/.build-cache
+          load: true
 
       - name: Run test recipes on host
         run: |
           docker-compose -f docker/recipes.test.yml \
-            up --build --exit-code-from=sut
+            up --exit-code-from=sut
+
+  uml-tests:
+    name: Run UML tests
+    needs:
+      - build
+    runs-on: ubuntu-latest
+    steps:
+      - name: Repository checkout
+        uses: actions/checkout@v2
+
+      - name: Setup Docker buildx
+        uses: docker/setup-buildx-action@v1
+
+      - name: Use cache
+        uses: actions/cache@v2
+        with:
+          path: /tmp/.build-cache
+          key: ${{ runner.os }}-docker-${{ github.sha }}
+
+      - name: Build Docker image
+        uses: docker/build-push-action@v2
+        with:
+          context: .
+          push: false
+          tags: debos
+          file: docker/Dockerfile
+          cache-from: type=local,src=/tmp/.build-cache
+          load: true
 
       - name: Run test recipes using UML backend
         run: |
           docker-compose -f docker/recipes-test-uml.yml \
```

Comment:
now that no builds are done in docker-hub anymore can we get rid of the docker-compose wrapping? (can be later though)

Reply:
yeah - i also want to move Dockerfile to the root and combine the readme files - added a TODO :-)